### ISSUE TITLE
Crash in TextureSystem::texture with automip=1 and TextureOpt::firstchannel>0, when sampling mip levels

### ISSUE THREAD

**Issue by ix-dcourtois** [OpenImageIO/oiio#2959]
Edit: I forgot to mention we're running on OIIO 2.1.15. I'll try to run the same test on a more recent version.
Edit2: Still happening with 2.2.13.1

**Describe the bug**
We have an OpenEXR file with lots of AOVs, and we noticed some crashes when we tried to read any of those AOVs with the `automip` option set to `1`

I recompiled OIIO in debug to try and understand the cause of the problem, and I think I narrowed it down to https://github.com/OpenImageIO/oiio/blob/4824fc7457b698bdd6b3264b9b1d72c46143aec4/src/libOpenImageIO/imagebuf.cpp#L2062

In our case, the ROI has `chbegin=10` and `chend=13`. The result of that line is that the ROI now has `chbegin=10` and `chend=3`, resulting in the next calls to `roi.chnnels()` to return `-7`, leading to crashes (access violation) in the `parallel_convert_image` call happening just after.

At first, I tried to simply comment out that line. It fixes the crashes. BUT then the mipmaps seems incorrect:
![mipmaps](https://user-images.githubusercontent.com/29434962/116780074-7922fc00-aa7a-11eb-9d4e-7cd06fd6fd3a.gif)
When we zoom out, the colors shift, and there's even a sort of black band appearing.
This is why I'm opening this issue instead of a pull request.

**To Reproduce**
Just compile the following sample, and put the attached .exr file ([aov_multiple00000.zip](https://github.com/OpenImageIO/oiio/files/6409409/aov_multiple00000.zip))
 next to it:
```cpp
#include <OpenImageIO/texture.h>

int main(int argc, char ** argv) {
    OIIO::ustring filename("aov_multiple00000.exr");

    OIIO::TextureSystem *texture_system = OIIO::TextureSystem::create();
    texture_system->attribute("automip", 1);

    OIIO::TextureOpt options;
    options.firstchannel = 10;

    float result[4] = { 0.0f, 0.0f, 0.0f, 0.0f };
    texture_system->texture(filename, options, 0.2, 0.2, 0.003, -0.003, 0.003, 0.003, 3, result);

    OIIO::TextureSystem::destroy(texture_system);
    return 0;
}
```

Edit: I did some further testing, because this crash seems too obvious for me being the only one to encounter it. And in our case, we have a negative value passed as `dtdx` (the `-0.003` value in the example above)
Making this value positive positive in that sample fixes the crash, so I thought those values needed to be always positive, but making it so also seems to screw the sampling (a bit like in the attached gif)

**Comment by ix-dcourtois** [comment]
Ok, I've found the correct fix. It lies in `ImageCacheFile::read_unmipped`.
The problem is here: https://github.com/OpenImageIO/oiio/blob/554f854f4a58c26410ce03f9b5d7be0d8a6bc7e2/src/libtexture/imagecache.cpp#L937 it sends an incorrect ROI which contains the initial `chbegin` and `chend` (which are `10` and `13`) to an image buffer that was created with only 3 channels. See https://github.com/OpenImageIO/oiio/blob/554f854f4a58c26410ce03f9b5d7be0d8a6bc7e2/src/libtexture/imagecache.cpp#L878-L879

I fixed this by changing line 937 to:
```cpp
    lores.get_pixels(ROI(0, tw, 0, th, 0, 1, 0, nchans), format, data);
```

This fixes the crash and mipmapping seems to work correctly.
Would it be ok if I open a pull request ?

**Comment by lgritzecho** [comment]
Yes, please do make a PR!



On May 3, 2021 2:24:31 a.m. PDT, Damien Courtois ***@***.***> wrote:
>Ok, I've found the correct fix. It lies in
>`ImageCacheFile::read_unmipped`.
>The problem is here:
>https://github.com/OpenImageIO/oiio/blob/554f854f4a58c26410ce03f9b5d7be0d8a6bc7e2/src/libtexture/imagecache.cpp#L937
>it sends an incorrect ROI which contains the initial `chbegin` and
>`chend` (which are `10` and `13`) to an image buffer that was created
>with only 3 channels. See
>https://github.com/OpenImageIO/oiio/blob/554f854f4a58c26410ce03f9b5d7be0d8a6bc7e2/src/libtexture/imagecache.cpp#L878-L879
>
>I fixed this by changing line 937 to:
>```cpp
>    lores.get_pixels(ROI(0, tw, 0, th, 0, 1, 0, nchans), format, data);
>```
>
>This fixes the crash and mipmapping seems to work correctly.
>Would it be ok if I open a pull request ?
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/OpenImageIO/oiio/issues/2959#issuecomment-831137585

-- 
Larry Gritz
***@***.***

**Comment by ix-dcourtois** [comment]
fixed by PR #2962
